File: packages/capacitor-actito-in-app-messaging/android/src/main/java/com/actito/iam/capacitor/ActitoInAppMessagingPlugin.kt
package com.actito.iam.capacitor

import android.os.Handler
import android.os.Looper
import com.actito.Actito
import com.actito.iam.ActitoInAppMessaging
import com.actito.iam.ktx.inAppMessaging
import com.actito.iam.models.ActitoInAppMessage
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

@CapacitorPlugin(name = "ActitoInAppMessagingPlugin")
public class ActitoInAppMessagingPlugin : Plugin(), ActitoInAppMessaging.MessageLifecycleListener {

    override fun load() {
        logger.hasDebugLoggingEnabled = Actito.options?.debugLoggingEnabled ?: false

        EventBroker.setup(this::notifyListeners)

        Actito.inAppMessaging().removeLifecycleListener(this)
        Actito.inAppMessaging().addLifecycleListener(this)
    }

    // region Actito In-App Messaging

    @PluginMethod
    public fun hasMessagesSuppressed(call: PluginCall) {
        try {
            call.resolve(
                JSObject().apply {
                    put("result", Actito.inAppMessaging().hasMessagesSuppressed)
                }
            )
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
        }
    }

    @PluginMethod
    public fun setMessagesSuppressed(call: PluginCall) {
        try {
            val suppressed = call.getBoolean("suppressed") ?: run {
                call.reject("Missing 'product' parameter.")
                return
            }

            val evaluateContext = call.getBoolean("evaluateContext") ?: false

            Actito.inAppMessaging().setMessagesSuppressed(suppressed, evaluateContext)

            call.resolve()
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
        }
    }

    // region ActitoInAppMessaging.MessageLifecycleListener

    override fun onMessagePresented(message: ActitoInAppMessage) {
        try {
            val data = JSObject().apply {
                put("message", message.toJson())
            }

            EventBroker.dispatchEvent("message_presented", data)
        } catch (e: Exception) {
            logger.error("Failed to emit the message_presented event.", e)
        }
    }

    override fun onMessageFinishedPresenting(message: ActitoInAppMessage) {
        try {
            val data = JSObject().apply {
                put("message", message.toJson())
            }

            EventBroker.dispatchEvent("message_finished_presenting", data)
        } catch (e: Exception) {
            logger.error("Failed to emit the message_finished_presenting event.", e)
        }
    }

    override fun onMessageFailedToPresent(message: ActitoInAppMessage) {
        try {
            val data = JSObject().apply {
                put("message", message.toJson())
            }

            EventBroker.dispatchEvent("message_failed_to_present", data)
        } catch (e: Exception) {
            logger.error("Failed to emit the message_failed_to_present event.", e)
        }
    }

    override fun onActionExecuted(message: ActitoInAppMessage, action: ActitoInAppMessage.Action) {
        try {
            val data = JSObject().apply {
                put("message", message.toJson())
                put("action", action.toJson())
            }

            EventBroker.dispatchEvent("action_executed", data)
        } catch (e: Exception) {
            logger.error("Failed to emit the action_executed event.", e)
        }
    }

    override fun onActionFailedToExecute(
        message: ActitoInAppMessage,
        action: ActitoInAppMessage.Action,
        error: Exception?
    ) {
        try {
            val data = JSObject().apply {
                put("message", message.toJson())
                put("action", action.toJson())
            }

            if (error != null) {
                data.put("error", error.message)
            }

            EventBroker.dispatchEvent("action_failed_to_execute", data)
        } catch (e: Exception) {
            logger.error("Failed to emit the action_failed_to_execute event.", e)
        }
    }

    // endregion

    // endregion

    public companion object {
        internal fun onMainThread(action: () -> Unit) = Handler(Looper.getMainLooper()).post(action)
    }
}

File: packages/capacitor-actito-in-app-messaging/ios/Sources/ActitoInAppMessagingPlugin/ActitoInAppMessagingPlugin.swift
import ActitoInAppMessagingKit
import ActitoKit
import Capacitor
import Foundation

@objc(ActitoInAppMessagingPlugin)
public class ActitoInAppMessagingPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "ActitoInAppMessagingPlugin" 
    public let jsName = "ActitoInAppMessagingPlugin" 
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "hasMessagesSuppressed", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setMessagesSuppressed", returnType: CAPPluginReturnPromise),
    ] 
    public override func load() {
        addApplicationLaunchListener()

        EventBroker.instance.setup { self.notifyListeners($0, data: $1) }
        Actito.shared.inAppMessaging().delegate = self
    }

    @objc func hasMessagesSuppressed(_ call: CAPPluginCall) {
        call.resolve([
            "result": Actito.shared.inAppMessaging().hasMessagesSuppressed
        ])
    }

    @objc func setMessagesSuppressed(_ call: CAPPluginCall) {
        guard let suppressed = call.getBool("suppressed") else {
            call.reject("Missing 'suppressed' parameter.")
            return
        }

        let evaluateContext = call.getBool("evaluateContext") ?? false

        Actito.shared.inAppMessaging().setMessagesSuppressed(suppressed, evaluateContext: evaluateContext)

        call.resolve()
    }
}

extension ActitoInAppMessagingPlugin: ActitoInAppMessagingDelegate {
    public func actito(_ actito: ActitoInAppMessaging, didPresentMessage message: ActitoInAppMessage) {
        do {
            let data = [
                "message": try message.toJson()
            ]

            EventBroker.instance.dispatchEvent("message_presented", data: data)
        } catch {
            logger.error("Failed to emit the message_presented event.", error: error)
        }
    }

    public func actito(_ actito: ActitoInAppMessaging, didFinishPresentingMessage message: ActitoInAppMessage) {
        do {
            let data = [
                "message": try message.toJson()
            ]

            EventBroker.instance.dispatchEvent("message_finished_presenting", data: data)
        } catch {
            logger.error("Failed to emit the message_finished_presenting event.", error: error)
        }
    }

    public func actito(_ actito: ActitoInAppMessaging, didFailToPresentMessage message: ActitoInAppMessage) {
        do {
            let data = [
                "message": try message.toJson()
            ]

            EventBroker.instance.dispatchEvent("message_failed_to_present", data: data)
        } catch {
            logger.error("Failed to emit the message_failed_to_present event.", error: error)
        }
    }

    public func actito(_ actito: ActitoInAppMessaging, didExecuteAction action: ActitoInAppMessage.Action, for message: ActitoInAppMessage) {
        do {
            let data = [
                "message": try message.toJson(),
                "action": try action.toJson(),
            ]

            EventBroker.instance.dispatchEvent("action_executed", data: data)
        } catch {
            logger.error("Failed to emit the action_executed event.", error: error)
        }
    }

    public func actito(_ actito: ActitoInAppMessaging, didFailToExecuteAction action: ActitoInAppMessage.Action, for message: ActitoInAppMessage, error: Error?) {
        do {
            var data: [String: Any] = [
                "message": try message.toJson(),
                "action": try action.toJson(),
            ]

            if let error = error {
                data["error"] = error.localizedDescription
            }

            EventBroker.instance.dispatchEvent("action_failed_to_execute", data: data)
        } catch {
            logger.error("Failed to emit the action_failed_to_execute event.", error: error)
        }
    }
}

extension ActitoInAppMessagingPlugin {
    private func addApplicationLaunchListener() {
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(didFinishLaunching),
            name: UIApplication.didFinishLaunchingNotification,
            object: nil
        )
    }

    private func removeApplicationLaunchListener() {
        NotificationCenter.default.removeObserver(
            self,
            name: UIApplication.didFinishLaunchingNotification,
            object: nil
        )
    }

    @objc private func didFinishLaunching() {
        removeApplicationLaunchListener()

        logger.hasDebugLoggingEnabled = Actito.shared.options?.debugLoggingEnabled ?? false
    }
}
